tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +224,286 @@ static bool HasCanonicalDimensionNumbers(
+    }
+    // Check that all ellipsis represent same rank of batch
+    if (batch_rank_vec.size() > 1) {
+      if (!std::equal(batch_rank_vec.begin() + 1, batch_rank_vec.end(),
+                      batch_rank_vec.begin())) {
+        return rewriter.notifyMatchFailure(
+            op, "Invalid Elipsis(ellipsis) within Einsum config!");
+      }
+    }
+
+    // Find all unique indices in the input and output
+    llvm::SmallSetVector<StringRef, 4> input_ind;
+    llvm::SmallSetVector<StringRef, 4> output_ind;
+    int num_of_input_operands = args.size();
+    for (const auto& loops_it : llvm::enumerate(all_loop_dim)) {
+      for (const StringRef& dim : loops_it.value()) {
+        if (loops_it.index() < max_operand_sz - 1) {
+          input_ind.insert(dim);
+        } else {
+          output_ind.insert(dim);
+        }
+      }
+    }
+
+    // Check for contraction/summation indices
+    llvm::SmallSetVector<StringRef, 4> reduction_axe =
+        FindSummationAxes(input_ind, output_ind);
+
+    // Find input/output values and types.
+    auto loc = op.getLoc();
+    ValueRange inputs = args;
+
+    // Setting up Initial buffer for Output Tensor
+    Value output;
+    auto dyn_sizes = ExtractDynamicEinsumSizes(
+        rewriter, loc, args.front(), args.back(), all_loop_dim[0],
+        all_loop_dim[1], all_loop_dim[2]);
+    output = GetInitTensor(rewriter, loc, result_ty, dyn_sizes);
+
+    // Create indexing maps.
+    // Create a 1:1 map from f:StrDimension -> affineDimension
+    int64_t nloops = input_ind.size();
+    DenseMap<StringRef, AffineExpr> str_affine_dim_umap;
+    for (auto it : llvm::enumerate(input_ind)) {
+      str_affine_dim_umap[it.value()] = rewriter.getAffineDimExpr(it.index());
+    }
+    // For each Operands, we generate the affine map for accessing pattern.
+    // Using knowledge from einsum_config/loop of each operand in string type,
+    // we can generate the equivalent information in affine type using
+    // "GetExprFromConfig". Then we can feed it into the affinemap builder
+    // where: map = mapBuilder(number_of_dims, num_constant, einsum_config,
+    // mlirCtx)
+    SmallVector<AffineMap, 4> maps;
+    for (const auto& loop_operand : all_loop_dim) {
+      AffineMap scalar_map = AffineMap::get(nloops, 0, rewriter.getContext());
+      auto inputExprs =
+          GetExprFromConfig(input_ind, loop_operand, str_affine_dim_umap);
+      maps.push_back(
+          AffineMap::get(nloops, 0, inputExprs, rewriter.getContext()));
+    }
+
+    // Creating Regions to be put into linalg.generic
+    // If has reduction axes, operation will become mulf+addf
+    // If no reduction is done, then operation will become just mulf
+    auto linalg_op = rewriter.create<linalg::GenericOp>(
+        loc, result_ty ? result_ty : TypeRange{}, inputs, output, maps,
+        GetEinsumLoopsAttrs(input_ind, reduction_axe),
+        [&](OpBuilder& nestedBuilder, Location nestedLoc, ValueRange args) {
```

Comment:
style nit: nested_builder, nested_loc.

I think we can use `b` and `nested_loc`.

Reply:
done